### DISCUSSION TITLE
Query Filtering Based Values Returning Bizarre Results

### DISCUSSION THREAD

**Discussion by RobertLbebber** [danielaparker/jsoncons#332] [Q&A] [answered]
I am trying to use `jsoncons::jsonpath::json_query` to returns an entry or a set of entries that fit the criteria defined within the query.
```
{
"shelves": [
    {
        "books" : [
                 {
                     "price" : 10,
                     "title" : "Fire on the Water"
                 }
        ]
    },
    {
          "books" : [
                      {
                          "price" : 5,
                          "title" : "House of Trees"
                      }
            ],
      },
      {
          "books" : [
                 {
                     "price" : 15,
                     "title" : "More Things"
                 }
           ]
      }
]
}
```
For instance, I want a query to return that returns the book with the title 'House of Trees'.
I have been trying to use the JSONPath format to query the information (https://goessner.net/articles/JsonPath/), like so:
_jsoncons::jsonpath::json_query(jsonDocument, "$..books[?(@.title='House of Trees']")_
and several other iterations of the same ideas, with no success. 

I have also tried to get the inner array and then do a filter on that inner array separately, to no success. 
Thoughts and Prays?

**Comment by danielaparker** [comment, marked as answer]
There are a few issues with the JSON you posted, I believe you meant to post
```json
{
    "shelves": [
        {
            "books": [
                {
                    "price": 10,
                    "title": "Fire on the Water"
                }
            ]
        },
        {
            "books": [
                {
                    "price": 5,
                    "title": "House of Trees"
                }
            ]
        },
        {
            "books": [
                {
                    "price": 15,
                    "title": "More Things"
                }
            ]
        }
    ]
}
```
Then the query should be
```
$.shelves[*].books[?(@.title == 'House of Trees')]
```
That should give you
```json
[{"price":5,"title":"House of Trees"}]
```
Note that you need the wildcard selector after 'shelves', otherwise `.books` will be evaluated against an array, and that's a type error; `.books` needs to be evaluated against each of the contained objects.

You could also use the query
```
$..[?(@.title == 'House of Trees')]
```

**Reply by RobertLbebber** [reply]
Hmm say for instance the shelves entries had a unique id like so:
```
     //...     
     {
            "uid": "2",  
            "books": [
                {
                    "price": 5,
                    "title": "House of Trees"
                }
            ]
        },
```

Given that we want to filter based on title, would we be able to recover the uid at that same time?
So effectively returning the shelves entry based on the book's title.

**Reply by RobertLbebber** [reply]
To solve this question I found this works:
___________________________________________[..........................Section A.......................] [ ............Section B.............]
`jsoncons::jmespath::search(json,  "shelves[?books[?title=='House of Trees'].{uid:uid,books:books})"`
This generates the subsequent json:
```
[
{
            "uid": "2",  
            "books": [
                {
                    "price": 5,
                    "title": "House of Trees"
                }
            ]
        }
]
```
Section A is the criteria to filter the shelves against, this is a nested criterion where the inner criteria returns true to the outer one if the book's title is "House of Trees". If the book title is false the entry will be dropped (filtered out).

Then in Section B the desired fields are plucked from the filtered json. In my case, I cared about both the uid and the books. But you can imagine there being other keys that this section would drop.

**Comment by danielaparker** [comment]
You could also use a jsoncons JSONPath extension, the parent operator `^` (first introduced by [JSONPath Plus](https://www.npmjs.com/package/jsonpath-plus)):
```
$.shelves[*].books[?(@.title == 'House of Trees')]^^
```
This returns the grandparent of the selected node:
```json
[{"books":[{"price":5,"title":"House of Trees"}],"uid":2}]
```
